utility/src/main/java/org/oppia/util/parser/LiTagHandler.kt

```diff
@@ -1,37 +1,171 @@
 package org.oppia.util.parser
 
+import android.content.Context
 import android.text.Editable
 import android.text.Html
 import android.text.Spannable
 import android.text.Spanned
-import android.text.style.BulletSpan
+import android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
+import android.text.Spanned.SPAN_MARK_MARK
+import android.text.style.LeadingMarginSpan
+import java.util.Stack
+import org.oppia.util.parser.LiTagHandler.ListTag
+import org.oppia.util.parser.StringUtils.LI_TAG
+import org.oppia.util.parser.StringUtils.OL_TAG
+import org.oppia.util.parser.StringUtils.UL_TAG
 import org.xml.sax.XMLReader
 
 /**
- * [Html.TagHandler] implementation that processes <li> tags and creates bullets.
+ * Called when the HTML parser reaches an opening or closing tag.
+ * We only handle list tags and ignore other tags.
  *
- * Reference: https://github.com/davidbilik/bullet-span-sample
+ * <ul> and <ol> tags are pushed to the [lists] stack and popped when the closing tag is reached.
+ *
+ * <li> tags are handled by the [ListTag] instance corresponding to the parent tag.
+ *
+ * Reference: https://github.com/daphliu/android-spannable-list-sample/tree/master/app/src/main/java/com/daphneliu/sample/listspansample
  */
-class LiTagHandler : Html.TagHandler {
+class LiTagHandler(private val context: Context) : Html.TagHandler {
+
+  private val lists = Stack<ListTag>()
+
+  override fun handleTag(opening: Boolean, tag: String, output: Editable, xmlReader: XMLReader) {
```

Comment:
I think we should add some resilience to cases when the HTML might be malformed (I'm not actually sure how the TagHandler is called in such cases). Can you investigate & add tests for the following cases?
- A string with ``<ol>`` and no closing tag; ditto for ``<li>`` and ``<ul>``
- A string that incorrectly mixes the lists, e.g. ``<li><ol></li></ol>``
- A string that closes one of the tags immediately, e.g. ``<li />`` or ``<ul />``

Assuming these cause issues, please make necessary changes here to avoid the app crashing in those cases.

Reply:
I tried something this sample https://bitbucket.org/Kuitsi/android-textview-html-list/src/master/app/src/main/assets/data.html.
![Screenshot_20200518-160018](https://user-images.githubusercontent.com/16301028/82203530-1300a000-9921-11ea-9510-5a274257ed1c.png)
![Screenshot_20200518-160026](https://user-images.githubusercontent.com/16301028/82203545-15fb9080-9921-11ea-8053-195b106795ea.png)

Reply:
We should have automated tests for such cases to make sure the behavior is what we expect (even if that behavior has weird visual behaviors).

Reply:
I have done some modifications to the code. Here is the updated 

> Thanks @veena14cs! Per the screenshot you added, the bullet spacing seems pretty large. Do we really want the spacing to be that large?

I have changed implementation and now the bullet spacing looks correct. Here is the updated screenshot.
![Screenshot_20200520-153824](https://user-images.githubusercontent.com/16301028/82435285-b4b6f700-9ab1-11ea-99c0-6ea9c7b3a81f.png)